deepfence_server/pkg/integration/teams/teams.go

```diff
@@ -35,35 +38,51 @@ func (t Teams) FormatMessage(message map[string]interface{}, position int) strin
 }
 
 func (t Teams) SendNotification(ctx context.Context, message string, extras map[string]interface{}) error {
+	t.client = utils.GetHttpClient()
+
 	var msg []map[string]interface{}
 	d := json.NewDecoder(strings.NewReader(message))
 	d.UseNumber()
 	if err := d.Decode(&msg); err != nil {
+		log.Info().Msgf("Failed to unmarshal message for teams: %v", err)
 		return err
 	}
 
-	startIndex := 0
-	endIndex := BatchSize
-	if endIndex > len(msg) {
-		endIndex = len(msg)
+	numRoutines := (len(msg) / 10)
+	if numRoutines == 0 {
+		numRoutines = 1
+	} else if numRoutines > 10 {
+		numRoutines = 10
 	}
-	if err := t.sendNotification(msg[startIndex:endIndex]); err != nil {
-		return err
+
+	senderChan := make(chan *Payload, 500)
+	var wg sync.WaitGroup
+	for i := 0; i < numRoutines; i++ {
+		wg.Add(1)
+		go t.Sender(senderChan, &wg)
 	}
+
+	defer func() {
+		close(senderChan)
+		wg.Wait()
+		log.Info().Msgf("Teams::SendNotification complete, numRoutines:%d", numRoutines)
+	}()
+
+	startIndex := 0
+	endIndex := BatchSize
+	t.enqueueNotification(msg[startIndex:endIndex], senderChan)
+
 	for endIndex < len(msg) {
 		startIndex = endIndex
 		endIndex += BatchSize
-		if endIndex > len(msg) {
-			endIndex = len(msg)
-		}
-		if err := t.sendNotification(msg[startIndex:endIndex]); err != nil {
-			return err
-		}
+		t.enqueueNotification(msg[startIndex:endIndex], senderChan)
 	}
 	return nil
 }
 
-func (t Teams) sendNotification(payloads []map[string]interface{}) error {
+func (t Teams) enqueueNotification(payloads []map[string]interface{},
+	senderChan chan *Payload) {
+
 	message := ""
 	for index, msgMap := range payloads {
 		message += t.FormatMessage(msgMap, index+1)
```

Comment:
Now I am curious - do we know how long this method takes?
This might be another bottleneck

Reply:
I don't think that is the case. In my testing, majority of time (> 99%) was going into the http send.